src/main/java/com/google/firebase/internal/TaskToApiFuture.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.internal;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.core.ApiFuture;
+import com.google.firebase.tasks.OnCompleteListener;
+import com.google.firebase.tasks.Task;
+import com.google.firebase.tasks.Tasks;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * An ApiFuture implementation that wraps a {@link Task}. This is an interim solution that enables
+ * us to expose Tasks as ApiFutures, until we fully remove the Task API.
+ *
+ * @param <T> Type of the result produced by this Future.
+ */
+public class TaskToApiFuture<T> implements ApiFuture<T> {
+
+  private final Task<T> task;
+
+  public TaskToApiFuture(Task<T> task) {
+    this.task = checkNotNull(task, "task must not be null");
+  }
+
+  @Override
+  public void addListener(final Runnable runnable, Executor executor) {
+    task.addOnCompleteListener(executor, new OnCompleteListener<T>() {
+      @Override
+      public void onComplete(Task<T> task) {
+        runnable.run();
+      }
+    });
+  }
+
+  @Override
+  public boolean cancel(boolean mayInterruptIfRunning) {
+    // Cannot be supported with Tasks
+    return false;
```

Comment:
I wonder if we should throw an exception instead since this will never do anything useful.  

Also, according to https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Future.html#cancel(boolean), "After this method returns, subsequent calls to isDone() will always return true." and I suspect that means that after calling cancel(), we shouldn't call their completion listener, etc. even if the actual work wasn't cancelled.

So if we don't throw, I think we need to do more work to follow the contract.

Reply:
I think the use of `cancel()` is commonplace enough, we need to do a bit better here than throwing an exception. I'm adding a boolean member to the class, and fixing the code so we correctly implement the contract of `isDone()`. 

The `cancel()` method only "attempts" to stop the task, and it is ok to not be able to stop the execution according to the contract. So I don't think we should try to stop any callbacks from firing.